go/vt/vtctl/workflow/server.go

```diff
@@ -1522,6 +1526,88 @@ func (s *Server) moveTablesCreate(ctx context.Context, req *vtctldatapb.MoveTabl
 	})
 }
 
+func (s *Server) validateRoutingRuleFlags(req *vtctldatapb.MoveTablesCreateRequest, mz *materializer) error {
+	if mz.IsMultiTenantMigration() {
+		switch {
+		case req.NoRoutingRules:
+			return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "cannot use --no-routing-rules in a multi-tenant migration")
+		case mz.isPartial:
+			return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "cannot run partial shard migration along with multi-tenant migration")
+		}
+	}
+	return nil
+}
+
+func (s *Server) setupInitialRoutingRules(ctx context.Context, req *vtctldatapb.MoveTablesCreateRequest, mz *materializer, tables []string, vschema *vschemapb.Keyspace) error {
+	if err := s.validateRoutingRuleFlags(req, mz); err != nil {
+		return err
+	}
+
+	const (
+		primaryType = ""
+		globalRoute
+	)
+	sourceKeyspace := req.SourceKeyspace
+	targetKeyspace := req.TargetKeyspace
+
+	if req.NoRoutingRules {
+		log.Warningf("Found --no-routing-rules flag, not creating routing rules for workflow %s.%s", targetKeyspace, req.Workflow)
+		return nil
+	}
+
+	// Save routing rules before vschema. If we save vschema first, and routing
+	// rules fails to save, we may generate duplicate table errors.
+	if mz.isPartial {
+		if err := createDefaultShardRoutingRules(mz.ctx, mz.ms, mz.ts); err != nil {
+			return err
+		}
+	}
+
+	if mz.IsMultiTenantMigration() {
+		log.Infof("Setting up keyspace routing rules for workflow %s.%s", targetKeyspace, req.Workflow)
+		var keyspaces []string
+		// Note that you can never point the target keyspace to the source keyspace in a multi-tenant migration
+		// since the target takes write traffic for all tenants!
+		keyspaces = append(keyspaces, sourceKeyspace)
+		if req.WorkflowOptions != nil && req.WorkflowOptions.SourceKeyspaceAlias != "" {
```

Comment:
Another place we'd benefit from:
```
req.GetWorkflowOptions().GetSourceKeyspaceAlias() != ""
```

As we're then protected from the case where req, req.WorkflowOptions, and req.WorkflowOptions.SourceKeyspaceAlias are nil.